### PR TITLE
feat(cq): xcpd initiation, post-processing and results storage

### PR DESCRIPTION
**Pull request by RamilGaripov**
refs. metriport/metriport-internal#1352

### Description

- Added XCPD trigger on patient create 
- Added storage for patient's CQ data containing successfully linked gateways and methods to create, update and delete it
- Added methods to retrieve XCPD results
- Added a function to create PD request objects
- Added an index for the `patient_discovery_result` table
- Fixed IHE Gateway routing
- Changed some of the IHE gateway schemas - THIS WILL RECEIVE ANOTHER UPDATE BEFORE RELEASE
- Added functionality to mock patient discovery responses (needs to be removed before merging)

### Testing

- Local
  - [x] Created patients locally and tested the flow with mocked PD and test payloads
 
- Staging
  - [ ] Run the same queries on staging if Mirth is set up to listen in the environment

### Release Plan
- [ ] Release NPM packages
- [ ] Merge this

### CODE REVIEW

File: packages/api/src/command/medical/patient/create-patient.ts [hunk 1]
```diff
@@ -44,12 +44,13 @@ export const createPatient = async (patient: PatientCreateCmd): Promise<Patient>
   // Intentionally asynchronous - it takes too long to perform
   cwCommands.patient.create(newPatient, facilityId).catch(processAsyncError(`cw.patient.create`));
 
-  // TODO: NEED CQ PATIENT DIRECTORY SEARCH
-  // const organization = await getOrganizationOrFail({ cxId });
+  const organization = await getOrganizationOrFail({ cxId });
+  const facility = await getFacilityOrFail({ cxId, id: facilityId });
```
Comment (Orta21) [inline packages/api/src/command/medical/patient/create-patient.ts:48]: Lets just create a variable on line 32 so we dont duplicate this call

File: packages/api/src/external/carequality/patient.ts [hunk 2]
```diff
@@ -1,15 +1,29 @@
...
+import {
+  PatientExternalUpdateCmd,
+  updateExternalData,
+} from "../../command/medical/patient/update-external-data";
 import { Organization } from "../../domain/medical/organization";
 import { Patient, PatientExternalData } from "../../domain/medical/patient";
 import { capture } from "../../shared/notifications";
-import { makeIheGatewayAPI } from "./api";
 import { toFHIR } from "../fhir/patient";
+// import { makeIheGatewayAPI } from "./api";
```
Comment (Orta21) [inline packages/api/src/external/carequality/patient.ts:16]: Leftover?

File: packages/api/src/external/carequality/patient.ts [hunk 3]
```diff
@@ -23,24 +37,74 @@ export async function discover(
   organization: Organization,
   facilityNPI: string
 ): Promise<void> {
-  let iheGateway: IHEGateway | undefined;
   try {
-    console.log(`IHE Gateway patient discovery - M patientId ${patient.id}`);
-
-    iheGateway = makeIheGatewayAPI();
+    const { id, cxId, data, eTag } = patient;
+    // const iheGateway = makeIheGatewayAPI();
```
Comment (Orta21) [inline packages/api/src/external/carequality/patient.ts:42]: Leftover

File: packages/api/src/external/carequality/patient.ts [hunk 4]
```diff
@@ -23,24 +37,74 @@ export async function discover(
...
       orgName: organization.data.name,
       orgOid: organization.oid,
     });
+    try {
+      // TODO: Uncomment this when testing is done
+      // await pTimeout(
+      //   iheGateway.startPatientDiscovery(iheGatewayRequest),
+      //   PATIENT_DISCOVERY_TIMEOUT.asMilliseconds(),
+      //   "startPatientDiscovery function timed out!"
+      // );
```
Comment (Orta21) [inline packages/api/src/external/carequality/patient.ts:65]: This is needed for testing in staging?

File: packages/api/src/external/carequality/patient.ts [hunk 5]
```diff
@@ -72,7 +136,7 @@ export async function handleDiscoverResponse(
     // WE SHOULD EXPECT THIS TO HAPPEN SO SHOULD WE ADD SENTRY MESSAGE?
     // DONT WANT TO BOMBARD ALERTS CHANNEL
     console.log(
```
Comment (Orta21) [inline packages/api/src/external/carequality/patient.ts:138]: Forgot to mention this https://github.com/metriport/metriport/pull/1302#discussion_r1424284557

File: packages/ihe-gateway-sdk/src/models/patient-discovery.ts [hunk 6]
```diff
@@ -19,12 +19,16 @@ export type PatientDiscoveryRequest = {
...
 };
 
 export const patientDiscoveryResponseSchema = baseResponseSchema.extend({
   patientMatch: z.boolean(),
-  xcpdHomeCommunityId: z.string().optional(),
+  xcpdHomeCommunityId: z.string(),
+  gateway: z.object({
+    oid: z.string(),
+    url: z.string(),
+  }),
```
Comment (Orta21) [inline packages/ihe-gateway-sdk/src/models/patient-discovery.ts:31]: This is already on the base schema

File: packages/ihe-gateway-sdk/src/models/patient-discovery.ts [hunk 7]
```diff
@@ -19,12 +19,16 @@ export type PatientDiscoveryRequest = {
   xcpdGateways: XCPDGateways;
   samlAttributes: SamlAttributes;
   patientResource: Patient;
-  principalCareProviderIds?: NPIStringArray;
+  principalCareProviderNPIs?: NPIStringArray;
```
Comment (Orta21) [inline packages/ihe-gateway-sdk/src/models/patient-discovery.ts:22]: Based on the schema this should be ids https://www.notion.so/metriport/XCPD-ITI-55-Interface-Outgoing-91b4e496b6b64ff6a041f55f3be5d1a5

File: packages/api/src/command/medical/patient-discovery-result/get-patient-discovery-result.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+export async function getPatientDiscoveryResults(
+  requestId: string
+): Promise<PatientDiscoveryResult[]> {
+  return await PatientDiscoveryResultModel.findAll({
+    where: {
+      requestId,
+      status: "success",
+    },
+  });
+}
```
Comment (Orta21) [inline packages/api/src/command/medical/patient-discovery-result/get-patient-discovery-result.ts:13]: Wondering if it would be beneficial to also receive failures that way we can log them or send them to sentry

**Review by Orta21 — COMMENTED** [review]
Added some comments, also since my code for uploading the responses to the db is in develop you can probably test the whole flow if you havent done so already

File: packages/ihe-gateway-sdk/src/models/patient-discovery.ts (hunk 7 above)
Comment (RamilGaripov) [reply packages/ihe-gateway-sdk/src/models/patient-discovery.ts:22]: Oh weird.. At some point, I merged develop, and it changed it to `...NPIs` . I'll revert this 👍

File: packages/ihe-gateway-sdk/src/models/patient-discovery.ts (hunk 6 above)
Comment (RamilGaripov) [reply packages/ihe-gateway-sdk/src/models/patient-discovery.ts:31]: Woops

File: packages/api/src/command/medical/patient/create-patient.ts (hunk 1 above)
Comment (RamilGaripov) [reply packages/api/src/command/medical/patient/create-patient.ts:48]: Ah, didn't notice we already checked the facility

File: packages/api/src/command/medical/patient-discovery-result/get-patient-discovery-result.ts (hunk 8 above)
Comment (leite08) [reply packages/api/src/command/medical/patient-discovery-result/get-patient-discovery-result.ts:13]: Isn't this the query on the table you created, @Orta21 ? If so, we need to add an index to the table, right?

File: packages/api/src/external/carequality/patient.ts (hunk 5 above)
Comment (RamilGaripov) [reply packages/api/src/external/carequality/patient.ts:138]: uh oh..

File: packages/api/src/external/carequality/patient.ts [hunk 9]
```diff
@@ -23,24 +37,74 @@ export async function discover(
...
+            context: createContext,
+            error,
+          },
+          level: "error",
+        });
+      }
+    }
+
+    // At this point, discovery results are stored in the database, so we can retrieve them
+    const discoveryResults = await getPatientDiscoveryResults(iheGatewayRequest.id);
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:93]: What's the logic here? Are we always waiting for the timeout, is `iheGateway.startPatientDiscovery` synchronous (as in, returns when the discovery is done)?

I thought the results of PD would go to lambdas

File: packages/api/src/external/carequality/patient.ts [hunk 10]
```diff
@@ -23,24 +37,74 @@ export async function discover(
...
+      id: patient.id,
+      cxId,
+      eTag,
+      COMMONWELL: data.externalData?.COMMONWELL ?? undefined,
+      CAREQUALITY: {
+        ...data.externalData?.CAREQUALITY,
+        gateways: linkedGateways,
+      },
+    };
+    await updateExternalData(updateData);
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:106]: I thought we wouldn't store these on the patient's table

File: packages/api/src/command/medical/patient/update-external-data.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+              ...patient.data.externalData?.CAREQUALITY,
+              ...patientUpdate.CAREQUALITY,
+            },
+          },
+        },
+      },
+      { transaction }
+    );
+  });
+};
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/command/medical/patient/update-external-data.ts:43]: ^ I think the link is wrong. It just points to the whole PR :(

File: packages/api/src/routes/medical/internal-cq.ts [hunk 12]
```diff
@@ -115,9 +116,32 @@ router.get(
...
+//   asyncHandler(async (req: Request, res: Response) => {
+//     const organization = await getOrganizationOrFail({ cxId });
+//     const patient = await getPatientOrFail({ id: patientId, cxId });
+
+//     await cqCommands.patient
+//       .discover(patient, organization, facilityNpi)
+//       .catch(processAsyncError(`cq.patient.discover`));
+//     return res.sendStatus(httpStatus.OK);
+//   })
+// );
```
Comment (Orta21) [inline packages/api/src/routes/medical/internal-cq.ts:138]: Leftover?

File: packages/api/src/routes/medical/internal-cq.ts [hunk 13]
```diff
@@ -115,9 +116,32 @@ router.get(
...
 
-    await handlePatientDiscoveryResponse(patientDiscovery);
+router.post(
+  "/patient-discovery/response-trigger",
+  asyncHandler(async (req: Request, res: Response) => {
+    const iheGateway = makeIheGatewayAPI();
+    iheGateway.startPatientDiscovery();
 
     return res.sendStatus(httpStatus.OK);
   })
```
Comment (Orta21) [inline packages/api/src/routes/medical/internal-cq.ts:147]: Is this for testing? Leftover?

File: packages/api/src/command/medical/patient-discovery-result/get-patient-discovery-result.ts (hunk 8 above)
Comment (Orta21) [reply packages/api/src/command/medical/patient-discovery-result/get-patient-discovery-result.ts:13]: @RamilGaripov whats the final verdict for this?

File: packages/api/src/external/carequality/patient.ts [hunk 14]
```diff
@@ -1,48 +1,86 @@
...
-      facilityNPI: facilityNPI,
+      xcpdGateways,
+      facilityNPI,
       orgName: organization.data.name,
       orgOid: organization.oid,
     });
 
-    iheGateway.startPatientDiscovery(iheGatewayRequest);
+    // Intentionally asynchronous - we will be checking for the results in the database
+    iheGateway.startPatientDiscovery([pdRequest]);
```
Comment (Orta21) [inline packages/api/src/external/carequality/patient.ts:56]: Whats going to be the release plan for this? Because AFAIK we wont be able to merge this into prod until the infra is setup

File: packages/api/src/external/carequality/mock-patient-discovery.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
+// THIS FILE SHOULD BE REMOVED WHEN TESTING IS DONE
```
Comment (jonahkaye) [inline packages/api/src/external/carequality/mock-patient-discovery.ts:1]: maybe dont remove it and just make it an integration test?

File: packages/api/src/models/medical/cq-patient-data.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
+import { DataTypes, Sequelize } from "sequelize";
+import { CQData, PatientCQData } from "../../domain/medical/cq-patient-data";
+import { BaseModel, ModelSetup } from "../_default";
+
+export class PatientCQDataModel extends BaseModel<PatientCQDataModel> implements PatientCQData {
```
Comment (jonahkaye) [inline packages/api/src/models/medical/cq-patient-data.ts:5]: OOC, whats the reason for a separate data model for CQ data as opposed to just adding it to the data in the current Patient Data model?

Comment (RamilGaripov) [reply packages/api/src/models/medical/cq-patient-data.ts:5]: So we've discussed this with the team, and the reason for this is that we're gonna have too much CQ data to be storing on the existing patient table. We're expecting up to 20-30 CQ links per patient, with 4 attributes for each for now. In the future, we might be adding even more stuff on there. Not only that, the patient update uses a table lock, and we want to avoid scenarios with conflicting updates happening at the same time.

File: packages/api/src/external/carequality/patient.ts [hunk 17]
```diff
@@ -1,48 +1,86 @@
...
+
+    const raceControl: RaceControl = { isRaceInProgress: true };
+    // Run the patient discovery until it either times out, or all the results are in the database
+    const raceResult = await Promise.race([
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:60]: TIL 🤩

File: packages/api/src/external/carequality/patient.ts [hunk 18]
```diff
@@ -1,48 +1,86 @@
...
+      startTimeout(),
+      checkNumberOfResults(raceControl, pdRequest.id, pdRequest.xcpdGateways.length), // TODO: RAISE A FOLLOW UP - set up an event listener for Mirth
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:62]: Is there a ticket for this TODO?

File: packages/api/src/external/carequality/patient.ts [hunk 19]
```diff
@@ -56,23 +94,72 @@ export async function discover(
...
+    }, timeout);
+  });
+}
+
+function checkNumberOfResults(
+  raceControl: RaceControl,
+  requestId: string,
+  numberOfGateways: number
+): Promise<string> {
+  return new Promise(resolve => {
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:145]: Why new promise here? If we declare the function `async` and return a function it will be a promise.

File: packages/api/src/external/carequality/patient.ts [hunk 20]
```diff
@@ -56,23 +94,72 @@ export async function discover(
...
+    const checkAndResolve = () => {
+      isPDComplete(requestId, numberOfGateways).then(isComplete => {
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:147]: NABD if really needed, but we prefer/use await/async across the code and avoid `.then()`

File: packages/api/src/external/carequality/patient.ts [hunk 21]
```diff
@@ -56,23 +94,72 @@ export async function discover(
...
+        if (isComplete) {
+          const msg = `Patient discovery results came back in full (${numberOfGateways} gateways)`;
+          resolve(msg);
+          raceControl.isRaceInProgress = false;
+        } else {
+          setTimeout(checkAndResolve, CHECK_DB_INTERVAL.asMilliseconds());
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:153]: if we move to await based can use `await sleep()`

File: packages/api/src/external/carequality/patient.ts [hunk 22]
```diff
@@ -56,23 +94,72 @@ export async function discover(
...
+      });
+    };
+
+    checkAndResolve();
+  });
+}
+
+async function isPDComplete(requestId: string, numGatewaysInRequest: number) {
+  const pdResultCount = await getPatientDiscoveryResultCount(requestId);
+  return pdResultCount === numGatewaysInRequest;
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:164]: `>=`, its safer

File: packages/api/src/external/carequality/patient.ts [hunk 23]
```diff
@@ -1,48 +1,86 @@
...
+      facilityNPI,
       orgName: organization.data.name,
       orgOid: organization.oid,
     });
 
-    iheGateway.startPatientDiscovery(iheGatewayRequest);
+    // Intentionally asynchronous - we will be checking for the results in the database
+    iheGateway.startPatientDiscovery([pdRequest]);
+
+    const raceControl: RaceControl = { isRaceInProgress: true };
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:58]: What are we using this for? I only see this being set, not read.

File: packages/api/src/external/carequality/patient.ts [hunk 24]
```diff
@@ -1,48 +1,86 @@
...
+        apiType: Product.medical,
+        numberGateways: xcpdGateways.length,
+        numberLinkedGateways: discoveryResults.length,
+      },
+    });
   } catch (err) {
-    const msg = `Failure while starting patient discovery for ${patient.id} @ IHE Gateway`;
+    const msg = `Failed to carry out patient discovery for ${patient.id}`;
     console.error(msg, err);
     capture.message(msg, {
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:85]: I know it was already there, but now we can use `capture.error(msg, ...)` - where `msg` is a string`, so we don't need to set `level: "error"`.

File: packages/api/src/external/carequality/patient.ts [hunk 25]
```diff
@@ -56,23 +94,72 @@ export async function discover(
...
+    return;
   }
+  const { id, cxId } = patient;
+
+  const cqLinks = buildCQLinks(pdResults);
+  const pdCreate: PatientCQDataCreate = {
+    id,
+    cxId,
+    data: { links: cqLinks },
+  };
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:112]: super nit: this doesn't need to execute if `cqLinks` is empty, so could just set this as parameter of `createOrUpdatePatientCQData` w/o creating a const.

File: packages/api/src/external/carequality/patient.ts [hunk 26]
```diff
@@ -56,23 +94,72 @@ export async function discover(
...
+    if (!id || !system) return [];
+    return {
+      patientId: id,
+      systemId: system,
+      ...pd.data.gateway,
+    };
+  });
+}
+
+function startTimeout() {
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:130]: nit: startTimeout gives the impression it's triggering the timeout; maybe something like `checkTimeout`, `controlDuration`?

File: packages/api/src/external/carequality/patient.ts [hunk 27]
```diff
@@ -56,23 +94,72 @@ export async function discover(
...
+  const timeout = PATIENT_DISCOVERY_TIMEOUT.asMilliseconds();
+  return new Promise<string>(resolve => {
+    setTimeout(() => {
+      const msg = `Patient discovery reached timeout after ${timeout} ms`;
+      resolve(msg);
+    }, timeout);
+  });
```
Comment (leite08) [inline packages/api/src/external/carequality/patient.ts:137]: ditto: why not async/await and use `sleep()`?

File: packages/api/src/routes/medical/internal-cq.ts (hunk 12 above)
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:138]: left-over?

File: packages/api/src/routes/medical/internal-cq.ts [hunk 28]
```diff
@@ -115,9 +116,32 @@ router.get(
 router.post(
   "/patient-discovery/response",
   asyncHandler(async (req: Request, res: Response) => {
-    const patientDiscovery = patientDiscoveryResponseSchema.parse(req.body);
+    console.log("Received patient discovery response from IHE Gateway", req.body);
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:119]: left-over?

File: packages/api/src/routes/medical/internal-cq.ts [hunk 29]
```diff
@@ -115,9 +116,32 @@ router.get(
...
+//     return res.sendStatus(httpStatus.OK);
+//   })
+// );
 
-    await handlePatientDiscoveryResponse(patientDiscovery);
+router.post(
+  "/patient-discovery/response-trigger",
+  asyncHandler(async (req: Request, res: Response) => {
+    const iheGateway = makeIheGatewayAPI();
+    iheGateway.startPatientDiscovery();
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:144]: This is odd or I might be missing something:
1. why `startPatientDiscovery` here if it's triggered upon Patient create?
2. if we're to start a PD here, wouldn't it make sense to have just `POST /patient-discovery`?

File: packages/ihe-gateway-sdk/src/client/ihe-gateway.ts [hunk 30]
```diff
@@ -41,11 +41,15 @@ export class IHEGateway {
...
    * https://profiles.ihe.net/ITI/TF/Volume2/ITI-55.html
    *
-   * @param patientDiscoveryRequest A patient discovery transaction request to Ihe Gateway.
+   * @param pdRequest A patient discovery transaction request to IHE Gateway.
    *
    */
-  async startPatientDiscovery(patientDiscoveryRequest: PatientDiscoveryRequest): Promise<void> {
-    await this.api.post(IHEGateway.PATIENT_DISCOVERY_ENDPOINT, patientDiscoveryRequest);
+  async startPatientDiscovery(pdRequest: PatientDiscoveryRequest[]): Promise<void> {
+    await this.api.post(IHEGateway.PATIENT_DISCOVERY_ENDPOINT, pdRequest, {
```
Comment (leite08) [inline packages/ihe-gateway-sdk/src/client/ihe-gateway.ts:48]: Ideally we would make input/outputs of the API always a JSON object, not an array. It allows to add more properties later - and we prob wouldn't need to set the JSON header. :)

**Review by leite08 — COMMENTED** [review]
In addition to the comments below, also consider these, please:
- update the PR title (not storing on patient external data any longer);
- missing link to upstream PR? It doesn't point to `develop`.

File: packages/api/src/domain/medical/cq-patient-data.ts [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+};
+
+export type CQDataWithCxId = {
+  cxId: string;
+  data: CQData;
+};
+
+export interface PatientCQDataCreate extends BaseDomainCreate, CQDataWithCxId {}
+
+export interface PatientCQData extends BaseDomain, PatientCQDataCreate {}
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/domain/medical/cq-patient-data.ts:27]: I think this is subjective and depends on how you associate this data. In my mind, it makes more sense to keep `patient` first, since this is very tightly coupled with the patient - this is why it was initially supposed to be in the patient table. And in the future, for XCA requests, we will be querying this table based on the patient ID because this data is patient-centric. 

What do you think?

File: packages/api/src/command/medical/cq-patient-data/create-cq-data.ts
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/command/medical/cq-patient-data/create-cq-data.ts:1]: But it's so nice to have all the Sequelize models in one place, no?

File: packages/api/src/domain/medical/cq-patient-data.ts (hunk 31 above)
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api/src/domain/medical/cq-patient-data.ts:27]: _(sorry for the long answer, but I think its important to share my perspective from diff angles)_

> ...it makes more sense to keep patient first, since this is very tightly coupled with the patient - this is why it was initially supposed to be in the patient table

I see what you mean! And I would 💯 agree with you it this was part of the Metriport logic... but this is specific to CQ. _(it might look like now we're tied to CQ, but that was the case 1 year ago w/ CW, and more HIEs will come)_

It was only originally suggested to store on the patient table b/c that's how CW came to be. And CW was stored on the patient table only b/c of development speed; since the beginning I've been promoting more separation between HIE-specific code/logic from the main logic of the API.

See how easy it is to attach a new HIE to the patient's creation flow, just add another of this specific to CQ on patient create:

```ts
cwCommands.patient.create(...)
```

...and, ideally, this wouldn't even be there it we had event based architecture from the start ([early steps ongoing now](https://metriport.slack.com/archives/C04DMKE9DME/p1702675771527909)).

Had we just coded to get it done, we wouldn't have saved much time and the maintenance of the api from day 0 would have been much more expensive.

Another perspective: imagine in a few months time we (hopefully) have a lot of load on our API and we decide to split services:
- the Metriport API
- a CW integration specific service/app
- a CQ integration specific service/app

It would be easier to move stuff around if we have CQ-related code together and likewise, CW. See OSS' `src/external/commonwell` and `src/external/carequality`, just move the folder around.

And there's the confusion with stuff like `getPatientCQData` - this is super close to other things we do with based off the Patient entity - whereas `getCQPatientData` is clearer as to what it's related to.

The main point is just separation of concerns, and keeping our logic/domain separated to HIE specific.

File: packages/api/src/command/medical/cq-patient-data/create-cq-data.ts
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api/src/command/medical/cq-patient-data/create-cq-data.ts:1]: I was referring to the command/logic/domain specific to CQ, but now that you mention I went to take a look and I don't see a reason to keep them on the main `models` folder that would overweight the separation between Metriport logic and logic specific to HIEs.

If you're referring to easy to find a model, I think its much faster to `cmd+o` and type `models/` than it is to scroll through the folder/files on VSCode. Not to mention we usually know the model name, so `cmd+o` and start typing the name is a direct hit.

File: packages/api/src/external/carequality/patient.ts (hunk 23 above)
Comment (leite08) [reply packages/api/src/external/carequality/patient.ts:58]: I understand the need to pass it as a reference, but what is it used for? I didn't find anywhere reading this, so in theory its not used and could be removed?

File: packages/api/src/external/carequality/patient.ts (hunk 24 above)
Comment (leite08) [reply packages/api/src/external/carequality/patient.ts:85]: It was, until a couple of days ago. Things move fast around here! 😅

File: packages/api/src/domain/medical/cq-patient-data.ts (hunk 31 above)
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/domain/medical/cq-patient-data.ts:25]: Yea this makes a lot of sense from the separation of concerns pov. Thanks for the detailed write up :)

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts
Comment (leite08) [inline packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:1]: Isn't this already deployed in staging/production? If so, we should not make any changes to it, instead create a new migration to update this table.

File: packages/api/src/routes/medical/internal-cq.ts [hunk 32]
```diff
@@ -110,13 +112,15 @@ router.get(
...
+ * POST /internal/carequality/patient-discovery/response/response
  *
  * Receives a Patient Discovery response from the IHE Gateway
  */
 router.post(
   "/patient-discovery/response",
   asyncHandler(async (req: Request, res: Response) => {
+    const pdResponse = patientDiscoveryResponseSchema.parse(req.body);
+    await createPatientDiscoveryResult(pdResponse);
     await handleIHEResponse({
```
Comment (leite08) [inline packages/api/src/routes/medical/internal-cq.ts:124]: Doesn't it make sense to have `createPatientDiscoveryResult()` being called inside `handleIHEResponse()`?

Also, since we're adding a schema for the input, we could pass the typed to const to `handleIHEResponse` (`req.body` is `any`).

**Comment by leite08** [comment]
Also, I see the release plan proposes deleting the table and migration manually. Can we do this through DB migrations? We had issues in the past trying to rollback stuff manually, also for whoever ran it on their computer would need to execute these manually on their DB otherwise it would break to run this new version.

File: packages/api/src/external/carequality/command/cq-patient-data/update-cq-data.ts [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+import { CQPatientDataModel } from "../../models/cq-patient-data";
+import { getCQPatientDataOrFail } from "./get-cq-data";
+
+export type CQPatientDataUpdate = CQPatientDataCreate & BaseUpdateCmdWithCustomer;
+
+export async function updateCQPatientData(cqData: CQPatientDataUpdate) {
+  const { id, cxId } = cqData;
+
+  return executeOnDBTx(CQPatientDataModel.prototype, async transaction => {
+    const cqPatientData = await getCQPatientDataOrFail({
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/external/carequality/command/cq-patient-data/update-cq-data.ts:14]: Currently, this function will get called once after each XCPD for a patient (after all gateways are processed / timeout reached). So, technically, with the current flow, it shouldn't get multiple request at the same time. But potentially in the future we might be getting those if more places update cq data for a patient.

File: packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts
Comment (RamilGaripov) [reply packages/api/src/sequelize/migrations/2023-12-12_00_create-patient-discovery.ts:1]: Since this hasn't been used in either prod or staging, we might want to do what we did last time with `cq_directory_entry` table, and just wipe it on both DBs and recreate them (I have this outlined in the release plan). 

But it would actually be easier and less error prone to create a new migration file.

File: packages/api/src/external/carequality/patient.ts (hunk 24 above)
Comment (RamilGaripov) [reply packages/api/src/external/carequality/patient.ts]: 🦔 🦸

File: packages/api/src/command/medical/patient/delete-patient.ts [hunk 34]
```diff
@@ -32,8 +33,18 @@ export const deletePatient = async (
...
+        cwCommands.patient.remove(patient, facilityId),
+        fhirApi.deleteResource("Patient", patient.id),
+        cqCommands.patient.remove(patient),
+      ]);
+
+      results.forEach(result => {
+        if (result.status === "rejected") {
+          throw result.reason;
+        }
+      });
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/command/medical/patient/delete-patient.ts:46]: Kk thx, I fixed it

File: packages/api/src/command/medical/ihe-result/create-ihe-result.ts [hunk 35]
```diff
@@ -37,21 +37,15 @@ export async function handleIHEResponse({ type, response }: IHEResult): Promise<
...
+      await createPatientDiscoveryResult(response);
       return;
     }
     case IHEResultType.DOCUMENT_QUERY: {
       const hasError = operationOutcome?.issue && !response.documentReference?.length;
 
       await DocumentQueryResultModel.create({
         ...defaultPayload,
-        status: hasError ? "success" : "failure",
+        status: hasError ? "failure" : "success",
```
Comment (RamilGaripov) [inline packages/api/src/command/medical/ihe-result/create-ihe-result.ts:48]: I noticed this when I merged develop. Think it was backwards, so I fixed it here

File: packages/api/src/external/carequality/models/patient-discovery-result.ts
Comment (RamilGaripov) [inline packages/api/src/external/carequality/models/patient-discovery-result.ts:1]: moved without making any changes

File: packages/api/src/external/carequality/patient.ts (hunk 23 above)
Comment (RamilGaripov) [reply packages/api/src/external/carequality/patient.ts:62]: I'm passing it to the function on line 66, and it's being used as a check in a while loop [here](https://github.com/metriport/metriport/blob/0b69a7b3bf7cfcb6bb72d20ef2267183337b6a07/packages/api/src/external/carequality/patient.ts#L144)

File: packages/api/src/command/medical/ihe-result/create-ihe-result.ts (hunk 35 above)
Comment (leite08) [reply packages/api/src/command/medical/ihe-result/create-ihe-result.ts:48]: NABD, but it would be a good function to avoid rewriting this multiple times

File: packages/api/src/command/medical/patient/delete-patient.ts [hunk 36]
```diff
@@ -33,33 +36,28 @@ export const deletePatient = async (
     try {
       // TODO: #393 move to declarative, event-based integration
       // Synchronous bc it needs to run after the Patient is deleted (it needs patient data from the DB)
-      const results = await Promise.allSettled([
-        cwCommands.patient.remove(patient, facilityId),
-        fhirApi.deleteResource("Patient", patient.id),
-        cqCommands.patient.remove(patient),
+      await Promise.allSettled([
+        cwCommands.patient.remove(patient, facilityId).catch(err => {
+          if (err.response?.status !== 404) throw err;
```
Comment (leite08) [inline packages/api/src/command/medical/patient/delete-patient.ts:41]: For this throw to reach the outer try/catch we need to move from `Promise.allSettled` to `Promise.all`, right?

Comment (RamilGaripov) [reply packages/api/src/command/medical/patient/delete-patient.ts:41]: Oh thats right

File: packages/api/src/command/medical/ihe-result/create-ihe-result.ts [hunk 37]
```diff
@@ -41,24 +44,40 @@ export async function handleIHEResponse({ type, response }: IHEResult): Promise<
...
     case IHEResultType.DOCUMENT_RETRIEVAL: {
-      const hasError = operationOutcome?.issue && !response.documentReference?.length;
-
       await DocumentRetrievalResultModel.create({
         ...defaultPayload,
-        status: hasError ? "failure" : "success",
+        status: getIheResultStatus({
+          operationOutcome,
+          docRefLength: response.documentReference?.length,
+        }),
```
Comment (leite08) [inline packages/api/src/command/medical/ihe-result/create-ihe-result.ts:63]: Defs NABD at this point, but if you send another commit, could move these out of the switch and calculate it once.

File: packages/api/src/command/medical/ihe-result/create-ihe-result.ts [hunk 38]
```diff
@@ -41,24 +44,40 @@ export async function handleIHEResponse({ type, response }: IHEResult): Promise<
...
         data: response,
       });
       return;
     }
   }
 }
+
+export function getIheResultStatus({
```
Comment (leite08) [inline packages/api/src/command/medical/ihe-result/create-ihe-result.ts:71]: nit: move to domain?

**Review by leite08 — APPROVED** [review]
Approved w/ a couple of nits. 🚀

File: packages/api/src/command/medical/ihe-result/create-ihe-result.ts (hunk 38 above)
Comment (RamilGaripov) [reply packages/api/src/command/medical/ihe-result/create-ihe-result.ts]: done

File: packages/api/src/command/medical/ihe-result/create-ihe-result.ts (hunk 37 above)
Comment (RamilGaripov) [reply packages/api/src/command/medical/ihe-result/create-ihe-result.ts]: since `response` of type `PatientDiscoveryResponse` has no `documentRefernece` attribute, I would have to implement another type check before calling this function, which looks kinda ugly.. I'll it like this for now. I think it's likely this function won't be called from here in the future, but rather inside corresponding functions similar to `createPatientDiscoveryResult`
